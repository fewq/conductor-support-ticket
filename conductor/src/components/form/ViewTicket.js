/* eslint-disable no-useless-constructor */
import React, { Component } from "react";
import axios from "axios";
import { convertDateToString } from "./helper";
import { Link } from 'react-router-dom';

export default class TicketList extends Component {
    constructor(props) {
        super(props);
        const {ticket} = this.props.location.state;
        const displayDate = convertDateToString(ticket.dateOfCreation);
        this.state = {ticket: ticket, displayDate: displayDate};
        this.routeChange = this.routeChange.bind(this);
    }

    alert() {
        if(window.confirm('Are you sure you want to delete this ticket?')) {
            this.delete();
        }
    }

    routeChange() {
        let path = `./restricted`;
        this.props.history.push(path);
      }

    delete() {
        axios.get('http://localhost:4000/ticket/delete/' + this.state.ticket._id)
          .then(res => {
            console.log('deleted' + this.state.ticket._id);
            this.routeChange();

          })
          .catch(err => console.log(err))
      }

    renderTopics() {
        return this.state.ticket.topics.map((obj, i) => {
            return <span class="badge badge-pill badge-info"> {obj} </span>
        });
    }

    render() {
        
        return(
            <div className="container-fluid">
                <div>
                    <h3> {this.state.ticket.title} <span class="badge badge-pill badge-secondary"> {this.state.ticket.statusToClient} </span></h3>
                    <div className="d-flex justify-content-center">
                    { this.renderTopics() }
                    </div>
                    <div>
                        <p> {this.state.displayDate} </p>
                        <h4> Description </h4>
                        <p> {this.state.ticket.description} </p>
                    </div>
                </div>
                <Link to={"/edit/"+this.state.ticket._id} className="btn btn-light">Edit Description</Link>
                <button onClick={this.alert} className="btn btn-danger">Delete Ticket</button>
            </div>

        )
    }
}