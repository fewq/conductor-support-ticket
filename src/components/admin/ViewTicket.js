import React, { Component } from "react";
import axios from "axios";
import { 
  convertDateToString,
  renderLoading,
  renderTopics,
  renderStatusHistory,
  renderScreenshots, } from "../helper";
import { Link } from "react-router-dom";
import jwtDecode from "jwt-decode";

export default class TicketList extends Component {
  constructor(props) {
    super(props);
    const { ticket } = this.props.location.state;
    const displayDate = convertDateToString(ticket.dateOfCreation);
    const idToken = jwtDecode(localStorage.getItem("id_token"));
    const email = idToken.email;
    this.state = {
      ticket: ticket,
      displayDate: displayDate,
      statusUpdates: [],
      attendedBy: email,
      imgSources: [],
      screenshotsLoaded: false
    };
  }

  componentWillMount() {
    axios
      .get("http://localhost:4000/status/ticketid/" + this.state.ticket._id)
      .then(response => {
        console.log("querying for relevant updates");
        console.log("View response:", response);
        if (response.data != null) {
          this.setState({
            statusUpdates: response.data
          });
          console.log("updated state");
          console.log(this.state);
        }
      })
      .catch((error, res) => {
        console.log("no status update history for this ticket.");
        // console.log(error);
      });

    if (this.state.ticket.fileUpload.length != 0) {
      axios
        .get(
          "http://localhost:4000/ticket/view/" +
            this.state.ticket._id +
            "/fileupload"
        )
        // console.log(imgSources);
        .then(res => {
          console.log(res);
          this.setState({
            imgSources: res.data.map((obj, i) =>
              new Buffer(obj, "base64").toString("binary")
            ),
            screenshotsLoaded: true
          });
        })
        .catch((error, res) => {
          console.log("no status update history for this ticket.");
          // console.log(error);
        });
    }
  }

  // Upon confirmation to delete, the ticket will be deleted,
  // a status update will be created, and an email will be sent.

  // form modal pop up to be implemented later for inputting comments.
  alert = () => {
    let ticketId = this.state.ticket._id;
    if (window.confirm("Are you sure you want to delete this ticket?")) {
      axios
        .delete("http://localhost:4000/ticket/delete/" + ticketId)
        .then(res => {
          console.log("deleted" + ticketId);
        });

      const status = "Closed";
      var update = {
        statusToClient: status,
        statusToAdmin: status,
        ticketId: this.state.ticket._id,
        attendedBy: this.state.attendedBy,
        dateOfUpdate: new Date()
      };

      axios
        .post("http://localhost:4000/status/add", update)
        .then(res => {
          console.log("posting status update");
          console.log(res);
        })
        .catch(err => console.log(err));

      // set email content
      const title =
        "Closed Ticket: " + this.state.ticket.title + " (" + ticketId + ")";
      const message = "closed by " + this.state.attendedBy;
      const email = this.state.ticket.createdBy;
      const target = "client";
      const subject = "Your ticket has been closed";
      const link = ticketId;

      axios.post("/api/notify", {
        email,
        subject,
        title,
        status,
        message,
        target,
        link
      });

      // change route
      let path = "/dashboard";
      this.props.history.push(path);
    }
  };

  render() {
    return (
      <div className="container-fluid">
        <div>
          <h3> {this.state.ticket.title} </h3>
          <p>
            {" "}
            Submitted on: {this.state.displayDate}{" "}
            <span className="badge badge-secondary mr-2">
              {" "}
              {this.state.ticket.statusToClient}{" "}
            </span>{" "}
          </p>
          <div className="d-flex justify-content-center my-2">
            {renderTopics(this.state.ticket.topics)}
          </div>
          <div className="my-5 mb-2">
            <h4> Description </h4>
            <p> {this.state.ticket.description} </p>
          </div>
        
        {this.state.imgSources.length !== 0 && 
            (this.state.screenshotsLoaded ? (
              renderScreenshots(this.state.imgSources)
            ) : (
              renderLoading()
            ) 
          )}
        </div>
        
        {this.state.statusUpdates.length !== 0 && 
        renderStatusHistory(this.state.statusUpdates)}
        
        <div className="my-2">
          <Link
            to={"/update/" + this.state.ticket._id}
            className="btn btn-light"
          >
            Update Ticket Status
          </Link>
          <button onClick={this.alert} className="btn btn-danger">
            Delete Ticket
          </button>
        </div>
      </div>
    );
  }
}
