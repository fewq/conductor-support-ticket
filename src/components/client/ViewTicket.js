/* eslint-disable no-useless-constructor */
import React, { Component } from "react";
import axios from "axios";
import { convertDateToString } from "./helper";
import { Link } from 'react-router-dom';

export default class TicketList extends Component {
    constructor(props) {
        super(props);
        const {ticket} = this.props.location.state;
        const displayDate = convertDateToString(ticket.dateOfCreation);
        this.state = {
            ticket: ticket,
            displayDate: displayDate,
            acknowledged: false,
            statusUpdates: [],
        };
        
        this.routeChange = this.routeChange.bind(this);
    }

    componentWillMount() {
        axios.get('http://localhost:4000/status/ticketid/' + this.state.ticket._id)
            .then((response) => {
                console.log("querying for relevant updates");
                console.log(response);
                if (response.data != null) {
                    this.setState({ statusUpdates: response.data, acknowledged: true });
                    console.log("updated state");
                    console.log(this.state);
                }
            })
            .catch( (error , response) => {
                console.log(error);
            });
    }

    alert() {
        if(window.confirm('Are you sure you want to delete this ticket?')) {
            this.delete();
        }
    }

    routeChange() {
        let path = `./restricted`;
        this.props.history.push(path);
      }

    delete() {
        axios.get('http://localhost:4000/ticket/delete/' + this.state.ticket._id)
          .then(res => {
            console.log('deleted' + this.state.ticket._id);
            // this.routeChange();

          })
          .catch(err => console.log(err))
      }
    
    renderTopics() {
        if (this.state.statusUpdates.length != 0) {
            return this.state.ticket.topics.map((obj, i) => {
                return <span class="badge badge-pill badge-warning mr-2"> {obj} </span>
            });
        }
    }

    render() {
        return(
            <div className="container-fluid" >
                <div>
                    <h3> {this.state.ticket.title} </h3>
                    <p> Submitted on: {this.state.displayDate} <span class="badge badge-secondary mr-2"> {this.state.ticket.statusToClient} </span> </p>
                    <div className="d-flex justify-content-center my-2">
                        { this.renderTopics() }          
                    </div>
                    <div className="my-5 mb-2">
                        <h4> Description </h4>
                        <p> {this.state.ticket.description} </p>
                    </div>
                    
                </div>
                {this.state.statusUpdates.map((obj, i) => 
                    <div>
                        <p>Update History</p>
                        <table className="table table-striped text-white">
                            <thead>
                                <tr>
                                    <th>Date</th>
                                    <th>Attended By</th>
                                    <th>Comments</th>
                                    <th>Status</th>
                                </tr>
                            </thead>
                            <tbody>
                                <td>
                                    {obj.dateOfUpdate}
                                </td>
                                <td>
                                    {obj.attendedBy}
                                </td>
                                <td>
                                    {obj.comments}
                                </td>
                                <td>
                                    Some status field to be implemented later.
                                </td>
                            </tbody>
                        </table>
                    </div>
                    )
                }

                <div className="my-2">
                <Link to={"/update/"+this.state.ticket._id} className="btn btn-light">Edit Description</Link>
                <button onClick={this.alert} className="btn btn-danger">Delete Ticket</button>
                </div>
            </div>

        )
    }
}