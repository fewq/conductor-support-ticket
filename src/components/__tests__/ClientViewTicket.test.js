import React from 'react'
import ViewTicket from '../client/ViewTicket';
import axios from 'axios';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { render, wait, fireEvent, cleanup } from 'react-testing-library'
import {toBeInTheDocument, toHaveClass} from 'jest-dom'
expect.extend({toBeInTheDocument, toHaveClass})

afterEach(cleanup)

var mockTicketState = {
    state: {
        ticket: {
            id: 1
        }
    }
}

// asynchronous testing to be implemented

// describe("Renders ticket details correctly", () => {
//     afterEach(cleanup)
    

// })

// describe("Render Status Update History", () => {
//     afterEach(cleanup)
//     beforeEach(() => {
//         axios.get = jest.fn(() => Promise.resolve(mockTicket))
//     })

// })


test('snapshot testing', () => {
    const tree = renderer
      .create(<BrowserRouter>
      <ViewTicket location={mockTicketState} />
      </BrowserRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });