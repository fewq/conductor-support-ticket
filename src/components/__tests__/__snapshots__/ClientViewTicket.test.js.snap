// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing 1`] = `
<div
  className="container-fluid"
>
  <div>
    <h3>
       
       
    </h3>
    <p>
       Submitted on: 
      Invalid Date
       
      <span
        class="badge badge-secondary mr-2"
      >
         
         
      </span>
       
    </p>
    <div
      className="d-flex justify-content-center my-2"
    />
    <div
      className="my-5 mb-2"
    >
      <h4>
         Description 
      </h4>
      <p>
         
         
      </p>
    </div>
  </div>
  <div
    className="my-2"
  >
    <a
      className="btn btn-light"
      href="/update/undefined"
      onClick={[Function]}
    >
      Edit Description
    </a>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Delete Ticket
    </button>
  </div>
</div>
`;
